contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+
+The primary motivation for Storage vendors to adopt the interface is a desire to make their system available to as many users as possible with as little work as possible. The primary motivation for COs to adopt the interface is to invest in a mechanism that will enable their users to use as many different storage systems as possible. In addition, for Kubernetes, adopting CSI will have the added benefit of moving volume plugins out of tree, and enabling volume plugins to be containerized.
+
+### Links
+
+* [Container Storage Interface (CSI) Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a CSI compliant volume plugin (a CSI volume driver) in Kubernetes.
+
+## Goals
+
+* Define Kubernetes API for interacting with an arbitrary, third-party CSI volume drivers.
+* Define mechanism by which Kubernetes master and node components will securely communicate with an arbitrary, third-party CSI volume drivers.
+* Define mechanism by which Kubernetes master and node components will discover and register an arbitrary, third-party CSI volume driver deployed on Kubernetes.
+* Recommend packaging requirements for Kubernetes compatible, third-party CSI Volume drivers.
+* Recommend deployment process for Kubernetes compatible, third-party CSI Volume drivers on a Kubernetes cluster.
+
+## Non-Goals
+* Replace [Flex Volume plugin](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+  * The Flex volume plugin exists as an exec based mechanism to create “out-of-tree” volume plugins.
+  * Because Flex drivers exist and depend on the Flex interface, it will continue to be supported with a stable API.
+  * The CSI Volume plugin will co-exist with Flex volume plugin.
+
+## Design Overview
+
+To support CSI Compliant Volume plugins, a new in-tree CSI Volume plugin will be introduced in Kubernetes. This new volume plugin will be the mechanism by which Kubernetes users (application developers and cluster admins) interact with external CSI volume drivers.
+
+The `SetUp`/`TearDown` calls for the new in-tree CSI volume plugin will directly invoke `NodePublishVolume` and `NodeUnpublishVolume` CSI RPCs through a unix domain socket on the node machine.
+
+Provision/delete and attach/detach must be handled by some external component that monitors the Kubernetes API on behalf of a CSI volume driver and invokes the appropriate CSI RPCs against it.
+
+To simplify integration, the Kubernetes team will offer a container that captures all the Kubernetes specific logic and acts as a bridge between third-party containerized CSI volume driver and Kubernetes (each deployment of a volume plugin would have it’s own “bridge”).
```

Comment:
A gap I realized with flex volumes the other day while Mike and were prototyping - there's a lot of mount behavior that is difficult to recreate outside of the core Kubernetes code base - specifically tmpfs mounting and live updates.  I think in general we talked about being able to use CSI for injecting data into containers, beyond just storage interfaces (as a node level extension mechanism), and so anyone following that path has to recreate a lot of somewhat tricky code.

Would like to see a section on affordances for CSI implementors to not have to reinvent wheels for Secrets.

Reply:
Ya, ephemeral volumes is a use case we want to support. But we're going to punt on it for alpha, I have it on the list of items for beta though.